{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from nlabpy.parse.seq import parse_fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download and extract `hg19` assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 16K\r\n",
      "drwxrwsr-x 3 ilya lab 4.0K Feb  2 15:46 \u001b[0m\u001b[01;34m.\u001b[0m/\r\n",
      "drwxrwsr-x 7 ilya lab 4.0K Feb  2 15:43 \u001b[01;34m..\u001b[0m/\r\n",
      "drwxrwsr-x 3 ilya lab 4.0K Feb  2 14:27 \u001b[01;34mHomo_sapiens\u001b[0m/\r\n",
      "-rw-rw-r-- 1 ilya lab  217 Feb  2 15:46 illuminaClipping.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls -lah ../ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens/UCSC/hg19/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Genes\n",
      "Homo_sapiens/UCSC/hg19/Annotation/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/SmallRNA\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Variation\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/ChromInfo.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/SmallRNA/precursor.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Variation/snp131.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-08-30-21-45-18/Variation/snp132.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/refMrna.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/ChromInfo.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/SmallRNA/hairpin.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Variation/snp130.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Variation/snp131.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2011-01-27-18-25-49/Variation/snp132.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/ChromInfo.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/SmallRNA/precursor.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/snp131.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/snp135.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/snp137.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/snp138.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2013-03-06-11-23-03/Variation/snp132.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/ChromInfo.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/SmallRNA/precursor.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Variation/snp131.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Variation/snp135.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2012-03-09-03-24-41/Variation/snp132.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/SmallRNA/hairpin.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp142.txt.idx\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp142.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp142.vcf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp135.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp137.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2014-06-02-13-47-56/Variation/snp138.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Genes/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/README.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/SmallRNA/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/SmallRNA/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/SmallRNA/hairpin.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp142.txt.idx\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp142.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp142.vcf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp135.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp137.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2015-07-17-14-32-32/Variation/snp138.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-current\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/splice_sites-49.fa.2bpb\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/exon_coords.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/splice_sites-49.fa.vld\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/splice_sites-49.fa.idx\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_49/splice_sites-49.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/genes.gtf\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/refMrna.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/DATE.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/ChromInfo.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/refSeqSummary.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/cytoBand.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/refFlat.txt.gz\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/splice_sites-34.fa\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/splice_sites-34.fa.vld\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/exon_coords.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/splice_sites-34.fa.idx\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/splice_sites_34/splice_sites-34.fa.2bpb\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/refFlat.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/knownGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/knownToRefSeq.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/refGene.txt\n",
      "Homo_sapiens/UCSC/hg19/Annotation/Archives/archive-2010-09-27-22-25-17/kgXref.txt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.3.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.rev.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.rev.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.4.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.fa.fai\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Bowtie2Index/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chrM.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr20.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr19.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr1.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr7.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr16.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chrM.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr11.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr22.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr14.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr2.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr20.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr15.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM_profile.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr16.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr8.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr1.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr4.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr21.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr7.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr13.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chrX.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chrY.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr22.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr4.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr21.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chrY.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr21.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr9.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr6.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr3.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/sno.txt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr12.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr13.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr17.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr14.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr15.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr9.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr3.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr6.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/tRNA.txt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr19.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr11.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr6.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr18.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr9.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr17.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr5.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr18.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr4.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr8.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr2.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr2.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr18.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr17.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr5.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr8.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr12.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr20.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr10.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr19.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr12.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr5.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr13.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr1.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr15.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chrM.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr16.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chrX.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/miRBase/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/miRBase/knownMiR.gff3\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/miRBase/mature.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/miRBase/hairpin.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr10.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chrY.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chr14.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr22.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr7.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/FM.chr11.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/bwt.chrX.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr3.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/repeatMasker.txt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/MDSBowtieIndex/SA.chr10.idx\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa.amb\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa.sa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.amb\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.sa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.bwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.ann\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.rbwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.pac\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.rpac\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.5.x/genome.fa.rsa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa.amb\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa.sa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa.bwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa.ann\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa.pac\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/version0.6.0/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa.bwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa.ann\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa.pac\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BWAIndex/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chrY.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr21.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr5.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr3.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr2.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr6.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr16.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr20.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr15.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr12.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chrM.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr1.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr4.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr9.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr18.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr10.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr22.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr14.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chrX.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr11.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr13.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr19.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr8.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr17.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr7.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nhr\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nsq\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nsd\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nsi\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nin\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BlastDB/genome.fa.nog\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.2.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.rev.2.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.1.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.3.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.rev.1.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/BowtieIndex/genome.4.ebwt\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/genome_mfa.GA_conversion.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.3.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.rev.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.rev.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/GA_conversion/BS_GA.4.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/genome_mfa.CT_conversion.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.rev.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.4.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.2.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.rev.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.3.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/Bisulfite_Genome/CT_conversion/BS_CT.1.bt2\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/GenomeSize.xml\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/genome.dict\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/genome.fa.fai\n",
      "Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/genome.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/polyA.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/hum5SrDNA.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/chrM.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/polyC.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/humRibosomal.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/adapter_contam1.fa\n",
      "Homo_sapiens/UCSC/hg19/Sequence/AbundantSequences/phix.fa\n",
      "README.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "wget ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg19/Homo_sapiens_UCSC_hg19.tar.gz \\\n",
    "    --directory-prefix=../ref\n",
    "tar -xzvf ../ref/Homo_sapiens_UCSC_hg19.tar.gz -C ../ref\n",
    "rm ../ref/Homo_sapiens_UCSC_hg19.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build `GTF` dataframe from the `lncRNA` annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gtf_df(filename):\n",
    "    res = []\n",
    "    with open(filename, 'rt') as fi:\n",
    "        for line in fi:\n",
    "            fields = line.strip().split('\\t')\n",
    "            if fields[2] == 'exon':\n",
    "                rec = {}\n",
    "                idfields = fields[8].strip().split(';')\n",
    "                for idfield in idfields:\n",
    "                    if idfield:\n",
    "                        key, val = idfield.split()\n",
    "                        if key == 'transcript_id' or key == 'exon_number':\n",
    "                            rec.update({key: val.strip('\"')})\n",
    "\n",
    "                rec.update({'chr': fields[0],\n",
    "                          'start': int(fields[3]),\n",
    "                          'end': int(fields[4])})\n",
    "                res.append(rec)\n",
    "    return pd.DataFrame.from_records(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>start</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087174</td>\n",
       "      <td>1</td>\n",
       "      <td>137086878</td>\n",
       "      <td>TCONS_tallLncRNA_00203783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087502</td>\n",
       "      <td>2</td>\n",
       "      <td>137087403</td>\n",
       "      <td>TCONS_tallLncRNA_00203783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087222</td>\n",
       "      <td>1</td>\n",
       "      <td>137086878</td>\n",
       "      <td>TCONS_tallLncRNA_00203782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087502</td>\n",
       "      <td>2</td>\n",
       "      <td>137087381</td>\n",
       "      <td>TCONS_tallLncRNA_00203782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137086987</td>\n",
       "      <td>1</td>\n",
       "      <td>137086771</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087419</td>\n",
       "      <td>2</td>\n",
       "      <td>137087039</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087809</td>\n",
       "      <td>3</td>\n",
       "      <td>137087491</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087483</td>\n",
       "      <td>1</td>\n",
       "      <td>137087011</td>\n",
       "      <td>ENST00000392399.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        end exon_number      start              transcript_id\n",
       "0  chr2  137087174           1  137086878  TCONS_tallLncRNA_00203783\n",
       "1  chr2  137087502           2  137087403  TCONS_tallLncRNA_00203783\n",
       "2  chr2  137087222           1  137086878  TCONS_tallLncRNA_00203782\n",
       "3  chr2  137087502           2  137087381  TCONS_tallLncRNA_00203782\n",
       "4  chr2  137086987           1  137086771                   lnrCXCR4\n",
       "5  chr2  137087419           2  137087039                   lnrCXCR4\n",
       "6  chr2  137087809           3  137087491                   lnrCXCR4\n",
       "7  chr2  137087483           1  137087011          ENST00000392399.3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = gtf_df('../ref/lncRNA.gtf')\n",
    "gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract the sequence of the locus annotated in `lncRNA.gtf` plus 500 bp on each side"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>start</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087174</td>\n",
       "      <td>1</td>\n",
       "      <td>137086878</td>\n",
       "      <td>TCONS_tallLncRNA_00203783</td>\n",
       "      <td>CAAAACCTTTATTAGCATTTTGAACAGGTTCAGCTATTACTGAAAC...</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087502</td>\n",
       "      <td>2</td>\n",
       "      <td>137087403</td>\n",
       "      <td>TCONS_tallLncRNA_00203783</td>\n",
       "      <td>TTCACATTTTCGATGGTGTCCCTGGGCTCCACTTCAAGGGCAATGG...</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087222</td>\n",
       "      <td>1</td>\n",
       "      <td>137086878</td>\n",
       "      <td>TCONS_tallLncRNA_00203782</td>\n",
       "      <td>CAAAACCTTTATTAGCATTTTGAACAGGTTCAGCTATTACTGAAAC...</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087502</td>\n",
       "      <td>2</td>\n",
       "      <td>137087381</td>\n",
       "      <td>TCONS_tallLncRNA_00203782</td>\n",
       "      <td>TTCCTTATCCTGGATCTTGGCCTTCACATTTTCGATGGTGTCCCTG...</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137086987</td>\n",
       "      <td>1</td>\n",
       "      <td>137086771</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "      <td>GTTTATTCAAGGACAAGCAGTCTGAGAAATGGAGTTTTTGAAATAA...</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087419</td>\n",
       "      <td>2</td>\n",
       "      <td>137087039</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "      <td>GGTGCAGGGTTGACTCTTTCTGGATGTTGTAGTCAGAAAGAGTGCG...</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087809</td>\n",
       "      <td>3</td>\n",
       "      <td>137087491</td>\n",
       "      <td>lnrCXCR4</td>\n",
       "      <td>GTTAGCGGATATGACGAGGCTCCGAAACACCAGTCATGTCCAGCCA...</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr2</td>\n",
       "      <td>137087483</td>\n",
       "      <td>1</td>\n",
       "      <td>137087011</td>\n",
       "      <td>ENST00000392399.3</td>\n",
       "      <td>TAACAGCCACCCCTCAGGTGCAGGACCAGGTGCAGGGTTGACTCTT...</td>\n",
       "      <td>472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr        end exon_number      start              transcript_id  \\\n",
       "0  chr2  137087174           1  137086878  TCONS_tallLncRNA_00203783   \n",
       "1  chr2  137087502           2  137087403  TCONS_tallLncRNA_00203783   \n",
       "2  chr2  137087222           1  137086878  TCONS_tallLncRNA_00203782   \n",
       "3  chr2  137087502           2  137087381  TCONS_tallLncRNA_00203782   \n",
       "4  chr2  137086987           1  137086771                   lnrCXCR4   \n",
       "5  chr2  137087419           2  137087039                   lnrCXCR4   \n",
       "6  chr2  137087809           3  137087491                   lnrCXCR4   \n",
       "7  chr2  137087483           1  137087011          ENST00000392399.3   \n",
       "\n",
       "                                            sequence  seq_length  \n",
       "0  CAAAACCTTTATTAGCATTTTGAACAGGTTCAGCTATTACTGAAAC...         296  \n",
       "1  TTCACATTTTCGATGGTGTCCCTGGGCTCCACTTCAAGGGCAATGG...          99  \n",
       "2  CAAAACCTTTATTAGCATTTTGAACAGGTTCAGCTATTACTGAAAC...         344  \n",
       "3  TTCCTTATCCTGGATCTTGGCCTTCACATTTTCGATGGTGTCCCTG...         121  \n",
       "4  GTTTATTCAAGGACAAGCAGTCTGAGAAATGGAGTTTTTGAAATAA...         216  \n",
       "5  GGTGCAGGGTTGACTCTTTCTGGATGTTGTAGTCAGAAAGAGTGCG...         380  \n",
       "6  GTTAGCGGATATGACGAGGCTCCGAAACACCAGTCATGTCCAGCCA...         318  \n",
       "7  TAACAGCCACCCCTCAGGTGCAGGACCAGGTGCAGGGTTGACTCTT...         472  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parser = parse_fasta('../ref/Homo_sapiens/UCSC/hg19/Sequence/Chromosomes/chr2.fa')\n",
    "_, chr2 = next(parser)\n",
    "\n",
    "def get_seqs(rec):\n",
    "    return chr2[rec.start:rec.end]\n",
    "\n",
    "gtf['sequence'] = gtf.apply(get_seqs, axis=1)\n",
    "gtf['seq_length'] = gtf['end'] - gtf['start']\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lnrCXCR4\r\n",
      "AGGAGTTTCCAGGTGACCCCTGGAAGTCCCAGTGCATTGCAGTCTTAGCACATTGCTCgagaaggtgagggagaagaagagagaaatgaaagaaaatttccagatgaagaaaagacaggaaagacagaggaagaaaggagggagggagattgaataaaagaaagagggagaaggtgaagaaggaaagagagagagagaATATATATAACGCTTTTAGGTGTTACCTTTGATCAGGGCGATTGACCAAGGTCAGCTTTCTTCAACGTGTATTCAGAGGAGGGCTCATGTCCTATAAGGTATTCATTGGTGTTTTACGGGGGAAATTTTTAAAAAGTGGGGCAGGGAAATCCACTGGTCCCACCCATTTGGGAAGTGTTTGgttcagcaggtttctctggtgtagctcctctcagagcctttcgtaaactggagtgcattatggagctccaagatggggccatagtatacaatttctccttacattatttTATTGAGATATTGTTTATTCAAGGACAAGCAGTCTGAGAAATGGAGTTTTTGAAATAATGATCCAGGCCTTTCCTGCAACACTGAGCTGTTTCTTTCCTTTTCTTTTTTAACCATGCAACAAAACCTTTATTAGCATTTTGAACAGGTTCAGCTATTACTGAAACTTGTAATTTCTAAACTTAAGTTGGGGCAAATGGCTATACGGCAGAGTAATGCCATCACTGGGCACTGCGAATGCAAGACTGGAGAATTAACAGCCACCCCTCAGGTGCAGGACCAGGTGCAGGGTTGACTCTTTCTGGATGTTGTAGTCAGAAAGAGTGCGGCCATCTTCCAGCTGCTTGCCTGCAAAGATGAGCCTCTGCTGGTCGGGGCTGGGGGTGGGGGGGTGCCTTCTTTATCCTGGATCTTGGCCTTCACATTTTCCATGGTGTCACTGGGCTCCACTTCCAGGGTGATGGTCTTGCCAGTCAGGGTCTTCACGAAGATCTGCATACCACCTCTCAGACACAGGACCAGGTGCAGGGTCTACTCTTTCTGGATGTTATAGTCAGAATGAGTGCAGTCATCTTCCACCTGCTTGACTGCAAAGATGAGCCTCTGCTGGTCCGGGGTAATGCCTTCCTTATCCTGGATCTTGGCCTTCACATTTTCGATGGTGTCCCTGGGCTCCACTTCAAGGGCAATGGTCTTGCTGGTAAGGGTCTTCACGAAGATCTGCATTTTGACCTGTTAGCGGATATGACGAGGCTCCGAAACACCAGTCATGTCCAGCCACAGGGACACCACCACATACTCACCCAACAAAGCCAGTCATCCCTACCACTGAGCTATTTCTATGCGAGTTCTTCCCTTGGCCCTTAAGCTGGGATAAATCCCTGTCTTCATGCAAAGTTAGAGACATGATTAGATACAAGATCTACAATATTTGTGGATAAAAACCAAACAGTTCCTTAAGAAAACTACAACTATTTTTTTTGGCTGACACCAGAGTGAAATTTCCCCCATTTATCCCCCATCAGCCTTTGGTAGGAGCACAAAAGCTACGTGGCAGGGCACATTCCAGCACCATGCCCATGACACCAACTCTCGTTCATTCATTCCTTGACGTATTTACATTCAAACTCCGTCCTCGTTTGCTGCTGTGCTGCTGGTTCTGGCTCCAAGCACttctttccttcttttttttttgagacaaagtctcgctgtcacccaggctggagtgcagtggcgtgatctcagttcactgcaacctccgcctcctgggttcaagcgattctcctgtctcagcttcccgaatagctgggagtgggccaccacacctggctaatttttgtatttttagtagagagggagccatgttagccaggctggtcttgaactcctaacctcaggtgatccacccgccttggcctcccaaagtgctgggattacaggcttgagtcatcacacctggccTCCAAGCACTTCTTACTCTGTCCTCAGACTTACGTGCTCATGCCTGACTCCCATATCTTCAAAGTTGAAAATGTTCTGATTTGTTTTCTCG\r\n"
     ]
    }
   ],
   "source": [
    "fa_tpl = '>{}'\n",
    "with open('../ref/ref_locus.fa', 'wt') as fo:\n",
    "    header = fa_tpl.format('lnrCXCR4')\n",
    "    fo.write('{}\\n{}\\n'.format(header, chr2[gtf.start.min()-500:gtf.end.max()+500]))\n",
    "\n",
    "!head ../ref/ref_locus.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build `bowtie2` index for the locus reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"../ref/lncRNA_locus.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  ../ref/ref_locus.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 509\n",
      "Using parameters --bmax 382 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 382 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 2038 (target: 381)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 2038 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2039 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 507\n",
      "fchr[G]: 982\n",
      "fchr[T]: 1466\n",
      "fchr[$]: 2038\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4195178 bytes to primary EBWT file: ../ref/lncRNA_locus.1.bt2\n",
      "Wrote 516 bytes to secondary EBWT file: ../ref/lncRNA_locus.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2038\n",
      "    bwtLen: 2039\n",
      "    sz: 510\n",
      "    bwtSz: 510\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 128\n",
      "    offsSz: 512\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 11\n",
      "    numLines: 11\n",
      "    ebwtTotLen: 704\n",
      "    ebwtTotSz: 704\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 509\n",
      "Using parameters --bmax 382 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 382 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 2038 (target: 381)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 2038 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2039 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 507\n",
      "fchr[G]: 982\n",
      "fchr[T]: 1466\n",
      "fchr[$]: 2038\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4195178 bytes to primary EBWT file: ../ref/lncRNA_locus.rev.1.bt2\n",
      "Wrote 516 bytes to secondary EBWT file: ../ref/lncRNA_locus.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2038\n",
      "    bwtLen: 2039\n",
      "    sz: 510\n",
      "    bwtSz: 510\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 128\n",
      "    offsSz: 512\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 11\n",
      "    numLines: 11\n",
      "    ebwtTotLen: 704\n",
      "    ebwtTotSz: 704\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "bowtie2-build ../ref/ref_locus.fa ../ref/lncRNA_locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
